operator/controllers/runtimetaskgroup_controller.go

```diff
@@ -18,36 +18,326 @@ package controllers
+	// gets relevant settings from top level objects (or use defaults)
+	executionMode := operatorv1.OperationExecutionModeAuto
+	operationPaused := false
+
+	if operation != nil {
+		executionMode = operation.Spec.GetTypedOperationExecutionMode()
+		operationPaused = operation.Status.Paused
+	}
+
+	// Reconcile paused override from top level objects
+	r.reconcilePauseOverride(operationPaused, taskgroup)
+
+	// Handle deleted TaskGroup
+	if !taskgroup.DeletionTimestamp.IsZero() {
+		err = r.reconcileDelete(taskgroup)
+		if err != nil {
+			return
+		}
+	}
+	// Handle non-deleted TaskGroup
+	err = r.reconcileNormal(executionMode, taskgroup, log)
+	if err != nil {
+		return
+	}
+
+	// Always reconcile Phase at the end
+	r.reconcilePhase(taskgroup)
+
+	return
+}
+
+func (r *RuntimeTaskGroupReconciler) reconcilePauseOverride(operationPaused bool, taskgroup *operatorv1.RuntimeTaskGroup) {
+	// record paused override state change, if any
+	taskgrouppaused := operationPaused
+	recordPausedChange(r.recorder, taskgroup, taskgroup.Status.Paused, taskgrouppaused, "by top level objects")
+
+	// update status with paused override setting from top level objects
+	taskgroup.Status.Paused = taskgrouppaused
+}
+
+func (r *RuntimeTaskGroupReconciler) reconcileNormal(executionMode operatorv1.OperationExecutionMode, taskgroup *operatorv1.RuntimeTaskGroup, log logr.Logger) error {
+	// If the TaskGroup doesn't have finalizer, add it.
+	//if !util.Contains(taskgroup.Finalizers, operatorv1alpha1.TaskGroupFinalizer) {
+	//	taskgroup.Finalizers = append(taskgroup.Finalizers, operatorv1alpha1.TaskGroupFinalizer)
+	//}
+
+	// gets all the Node object matching the taskgroup.Spec.NodeSelector
+	// those are the Node where the task taskgroup.Spec.Template should be replicated (desired tasks)
+	nodes, err := listNodesBySelector(r.Client, &taskgroup.Spec.NodeSelector)
+	if err != nil {
+		return errors.Wrap(err, "failed to list nodes")
+	}
+
+	desired := filterNodes(nodes, taskgroup.Spec.GetTypedTaskGroupNodeFilter())
+
+	// gets all the Task objects matching the taskgroup.Spec.Selector.
+	// those are the current Task objects controlled by this deployment
+	current, err := listTasksBySelector(r.Client, &taskgroup.Spec.Selector)
+	if err != nil {
+		return errors.Wrap(err, "failed to list tasks")
+	}
+
+	log.Info("reconciling", "Nodes", len(desired), "Tasks", len(current.Items))
+
+	// match current and desired state, so the controller can determine what is necessary to do next
+	tasks := reconcileTasks(desired, current)
+
+	// update replica counters
+	taskgroup.Status.Nodes = int32(len(tasks.all))
+	taskgroup.Status.RunningNodes = int32(len(tasks.running))
+	taskgroup.Status.SucceededNodes = int32(len(tasks.completed))
+	taskgroup.Status.FailedNodes = int32(len(tasks.failed))
+	taskgroup.Status.InvalidNodes = int32(len(tasks.invalid))
```

Comment:
does it make sense to have these fields as `uint32`?
we are probably not having cases or handing of e.g. -1

Reply:
You are right, but I have checked and some of the replica counters in kubernetes (e.g Deployments) and they are int32. 
Considering that we are setting them using `len`, I prefer to keep them aligned

Reply:
ack